Pass Nodes : Add new nodes for creating, deleting and wedging passes

murraystevenson opened the pull request:
This introduces a trio of new nodes for creating, deleting and wedging passes. These provide core functionality which will be expanded upon in future PRs adding a new dedicated editor panel for creating and managing passes. As such, these nodes aren't super exciting in isolation, but are being added in advance of the upcoming UI in order to provide pipelines with existing internal implementations of render pass/layer management tools an opportunity to test migration of their tools to these new nodes and conventions in preparation for the future release...

This PR provides: 
- A new `Passes` node for appending names to the `pass:names` option.
- A new `DeletePasses` node for removing (or keeping) names from the `pass:names` option.
- A new `PassWedge` node for wedging tasks with the `pass` context variable populated for each wedge with a name from the `pass:names` option.
- Config changes to include `${pass}` in render output paths and when writing ass files from an ArnoldRender node
- `${pass}` presets for the NameSwitch and Spreadsheet selector plugs

johnhaddon commented on python/GafferScene/PassWedge.py:
```diff
@@ -0,0 +1,94 @@
+
+		self["__OptionQuery"] = GafferScene.OptionQuery()
+		self["__OptionQuery"]["scene"].setInput( self["__TimeWarp"]["out"] )
+		self["__OptionQuery"].addQuery( Gaffer.StringVectorDataPlug( "value", defaultValue = IECore.StringVectorData() ) )
```
I _think_ we can omit the "value" argument here and below, since that will be done by `addQuery()`.

murraystevenson replied:
I've improved that in 5f38896 and also simplified some redundant use of `defaultValue`.

johnhaddon commented on python/GafferSceneTest/PassWedgeTest.py:
```diff
@@ -0,0 +1,242 @@
+		script["disablePass"] = GafferScene.CustomOptions()
+		script["disablePass"]["in"].setInput( script["passes"]["out"] )
+		script["disablePass"]["options"].addChild( Gaffer.NameValuePlug( "", Gaffer.BoolPlug( "value", defaultValue = False ), True, "member1" ) )
+		script["disablePass"]["options"]["member1"]["name"].setValue( "pass:enabled" )
```
I think this line can be omitted if we replace "" with "pass:enabled" on the line above?

murraystevenson replied:
Simplified in 4572e7c.

murraystevenson commented on src/GafferScene/Passes.cpp:
```diff
@@ -0,0 +1,126 @@
+	}
+	else
+	{
+		result->members()[g_passNamesOptionName] = new IECore::StringVectorData( names );
```
Fixed in 23e7a46, assuming the `const_cast` is reasonable here.

murraystevenson commented on python/GafferSceneUI/PassesUI.py:
```diff
@@ -0,0 +1,67 @@
+	"description",
+	"""
+	Appends passes to the list of pass names stored in the `pass:names`
+	option in Gaffer's globals.
```
Good point! I've made the description less abstract in 2bc3715, and matched the language for DeletePasses in f73aba7.

murraystevenson commented on python/GafferSceneTest/PassesTest.py:
```diff
@@ -0,0 +1,197 @@
+		self.assertEqual( passes["out"].object( "/plane" ), IECoreScene.MeshPrimitive.createPlane( imath.Box2f( imath.V2f( -0.5 ), imath.V2f( 0.5 ) ) ) )
+		self.assertEqual( passes["out"].transform( "/plane" ), imath.M44f() )
+		self.assertEqual( passes["out"].bound( "/plane" ), imath.Box3f( imath.V3f( -0.5, -0.5, 0 ), imath.V3f( 0.5, 0.5, 0 ) ) )
+		self.assertEqual( passes["out"].childNames( "/plane" ), IECore.InternedStringVectorData() )
```
Improved in 5e7ecdf.

murraystevenson commented on src/GafferScene/DeletePasses.cpp:
```diff
@@ -0,0 +1,136 @@
+	IECore::CompoundObjectPtr result = new IECore::CompoundObject();
+	result->members() = inputGlobals->members();
+
+	if( auto data = result->member<IECore::StringVectorData>( g_passNamesOptionName ) )
```
Well spotted! I've removed the redundant `if` in 956b693.

murraystevenson commented on startup/gui/menus.py:
```diff
@@ -559,6 +559,10 @@ def __usdLightCreator( lightType ) :
 
+# Pass nodes
+
+nodeMenu.append( "/Pass/Passes", GafferScene.Passes )
```
Adjusted in acbfe28, f6b415f, 5f362c4.

murraystevenson commented on python/GafferSceneUI/PassWedgeUI.py:
```diff
@@ -0,0 +1,129 @@
+	task dispatch.
+
+	Specific passes can be disabled from wedging by setting the
+	`pass:enabled` option to `0` in contexts where that pass name is
```
Fixed in 9ba6b07.

murraystevenson commented on python/GafferSceneUI/PassWedgeUI.py:
```diff
@@ -0,0 +1,129 @@
+	of contexts, each time with a different value for the `pass`
+	context variable. The tasks can then be varied per pass by using
+	`${pass}` in an upstream Spreadsheet or NameSwitch's `selector`
+	or through use of a ContextQuery node or an expression.
```
I've taken a pass (no pun intended!) at improving the node description in 869c0d2.

murraystevenson commented:
Thanks John!  Feedback addressed with fixup comments inline. As far as the renaming goes, we've discussed it offline and there does seem merit in making these nodes present as less abstract and more render orientated. This should help with node discovery via the Tab menu, and make the `${pass}` context variable a little easier to understand. The `pass:` option prefixes have also been renamed to `renderPass:` for consistency, and they seem to fit pretty well with the existing `render:` prefixed options when viewed in the Scene Inspector.

The node, option and context variable rename occurs in the last six commits starting with a1a0858. It seemed a little easier to tackle this way rather than interleaving the rename within the other fixups, and I can go through and squash it all down and update the commit messages before merging...

johnhaddon commented:
Thanks Murray! LGTM. But perhaps it might be worth adding a small `> Tip` section in the RenderPassesUI that mentions that the list of passes is stored in the `renderPass:names` option? And was it intention that the menu item for DeleteRenderPasses omits the "Render" part? If you could squash down as you see fit and then merge that would be great...

murraystevenson commented:
Great! I've added the `> Tip` to RenderPassesUI mentioning `renderPass:names` and put the existing tip about names that already exist being moved to the end of the list under the `names` plug, which seems a better home for it.

Thanks for catching that "Render" missing from the DeleteRenderPasses menu item, that's also fixed as part of the squash. Merging.

johnhaddon commented:
Nice one, thanks Murray!
